stdlib/src/utils/static_tuple.mojo

```diff
@@ -254,3 +314,139 @@ struct StaticTuple[element_type: AnyRegType, size: Int](Sized):
+            i += 1
+
+    @always_inline
+    fn __init__(inout self, values: VariadicList[Self.ElementType]):
+        """Creates a tuple constant using the specified values.
+
+        Args:
+            values: The list of values.
+        """
+        _static_tuple_construction_checks[size]()
+        # size is checked via the method call below
+        self.array = _create_array_collection_elem[size, Self.ElementType](
+            values
+        )
+
+    @always_inline("nodebug")
+    fn __len__(self) -> Int:
+        """Returns the length of the array. This is a known constant value.
+
+        Returns:
+            The size of the list.
+        """
+        return size
+
+    @always_inline
+    fn __refitem__[
+        mutability: __mlir_type.i1, self_life: AnyLifetime[mutability].type
+    ](
+        self: Reference[Self, mutability, self_life].mlir_ref_type, index: Int
+    ) -> Reference[Self.ElementType, mutability, self_life]:
+        """Gets a reference to the array element at the given index.
+
+        Parameters:
+            mutability: The inferred mutability of the type.
+            self_life: The inferred lifetime of the variant type.
+
+        Args:
+            index: The index of the element.
+
+        Returns:
+            A mutable reference to the element at the given index.
+        """
+        debug_assert(-size <= index < size, "index must be within bounds")
+        var normalized_idx = index
+
+        if index < 0:
+            normalized_idx += Reference(self)[].size
+
+        var array_copy = Reference(Reference(self)[].array)
+        var ptr = __mlir_op.`pop.array.gep`(
+            array_copy.get_unsafe_pointer().address, normalized_idx.value
+        )
+        return AnyPointer(ptr).__refitem__()
+
+    @always_inline
+    fn __refitem__[
+        mutability: __mlir_type.i1,
+        self_life: AnyLifetime[mutability].type,
+        index: Int,
+    ](self: Reference[Self, mutability, self_life].mlir_ref_type) -> Reference[
+        Self.ElementType, mutability, self_life
+    ]:
+        """Gets a reference to the array element at the given index.
+
+        Parameters:
+            mutability: The inferred mutability of the type.
+            self_life: The inferred lifetime of the variant type.
+            index: The index of the element.
+
+        Returns:
+            A mutable reference to the element at the given index.
+        """
+        constrained[-size <= index < size, "index must be within bounds"]()
+        var normalized_idx = index
+
+        if index < 0:
+            normalized_idx += Reference(self)[].size
+
+        var array_copy = Reference[Self.type, mutability, self_life](
+            Reference(self)[].array
+        )
+        var ptr = __mlir_op.`pop.array.gep`(
+            array_copy.get_unsafe_pointer().address, normalized_idx.value
+        )
+        return AnyPointer(ptr).__refitem__()
+
+    fn as_ptr(inout self) -> AnyPointer[Self.ElementType]:
```

Comment:
**Question** Do we want to expose this public API of `as_ptr()` yet or have it come in a follow-on?  Of note, `String._as_ptr()` is private (and not `as_ptr()` despite it getting used in very-public ways by users and in the library code probably in some places.

This is non-blocking, but curious to get your thoughts.  It would also be good to think in the future about API naming here (do we like `as_ptr()`?).  It's also inherently unsafe since it returns an `AnyPointer` as you know.

CC: @ConnorGray

Reply:
Agreed, removed this method.

Reply:
I think keeping the feature but marking it private and subject to change with an underscore prefix would be better. I believe the ability to access the underlying storage of static_tuple was requested specifically and only added recently so it would be a shame to lose it.

Also, UnsafePointers or References to _array are created several times in the code.  UnsafePointer now converts to an immortal Reference with suffix []. So _as_ptr() and _as_ptr()[] could replace all these and _get_reference_unsafe could be removed.

Since creating an immortal Reference from an UnsafePointer is new maybe others more knowledgable than me need to confirm its use is a good idea.